cargo-miri/bin.rs

```diff
@@ -568,24 +602,36 @@ fn phase_cargo_rustc(args: env::Args) {
         _ => {},
     }
 
-    if !print && target_crate && is_runnable_crate() {
-        // This is the binary or test crate that we want to interpret under Miri.
-        // But we cannot run it here, as cargo invoked us as a compiler -- our stdin and stdout are not
-        // like we want them.
-        // Instead of compiling, we write JSON into the output file with all the relevant command-line flags
-        // and environment variables; this is used when cargo calls us again in the CARGO_TARGET_RUNNER phase.
-        let info = CrateRunInfo::collect(args);
-        let filename = out_filename("", "");
-        if verbose {
-            eprintln!("[cargo-miri rustc] writing run info to `{}`", filename.display());
+    if !print && is_runnable_crate() {
+        macro_rules! store {
+            ($info:expr) => {
+                let info = $info;
+                let filename = out_filename("", "");
+                if verbose {
+                    eprintln!("[cargo-miri rustc] writing run info to `{}`", filename.display());
+                }
+                info.store(&filename);
+                // For Windows, do the same thing again with `.exe` appended to the filename.
+                // (Need to do this here as cargo moves that "binary" to a different place before
+                // running it.)
+                info.store(&out_filename("", ".exe"));
+                return;
+            };
+        }
+        if target_crate {
+            // This is the binary or test crate that we want to interpret under Miri.
+            // But we cannot run it here, as cargo invoked us as a compiler -- our
+            // stdin and stdout are not like we want them.
+            // Instead of compiling, we write JSON into the output file with all the
+            // relevant command-line flags and environment variables; this is used when
+            // cargo calls us again in the CARGO_TARGET_RUNNER phase.
+            store!(CrateRunInfo::collect(args));
+        } else if ArgFlagValueIter::new("--extern").any(|krate| krate == "proc_macro") {
```

Comment:
I think instead of putting this into the same `if` arm as the "proper" runnable case, it might make more sense to make these separate conditions. Like, first `if !print && target_crate && is_runnable_crate()`, and then later `if is_runnable_crate () && ArgFlagValueIter::new("--extern").any(|krate| krate == "proc_macro")`. The two branches have barely anything in common anyway.